### PR TITLE
[SPARK-32479][PYSPARK] Fix the slicing logic in createDataFrame when converting pandas dataframe to arrow table

### PR DESCRIPTION
**Pull request by liangz1**
### What changes were proposed in this pull request?

In `spark.createDataFrame(pdf: pd.DataFrame)`, the pdf will be sliced to `num_slices=defaultParallelism` slices before converting to arrow batches. However, the current logic may result in fewer partitions than specified. See the next section for details.

This PR replaced the slicing logic with the silimar logic as in [ParallelCollectionRDD.scala#L125](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rdd/ParallelCollectionRDD.scala#L125)
~~~python
# replace conversion.py#L417-418
# step = -(-len(pdf) // self.sparkContext.defaultParallelism)  # round int up
# pdf_slices = (pdf.iloc[start:start + step] for start in xrange(0, len(pdf), step))
pdf_slices = (pdf.iloc[i * length // num_slices: (i + 1) * length // num_slices] for i in xrange(0, num_slices))
~~~

#### Problem
Currently, in [conversion.py#L418](https://github.com/databricks/runtime/blob/84a952313ae73e3df32f065eb00cc0bcb024af14/python/pyspark/sql/pandas/conversion.py#L418|https://github.com/databricks/runtime/blob/84a952313ae73e3df32f065eb00cc0bcb024af14/python/pyspark/sql/pandas/conversion.py#L418), the slicing logic may result in fewer partitions than specified.
###### Example:

Assume:
```
length = 100 -> [0, 1, ..., 99]
num_slices = 99 = self.sparkContext.defaultParallelism
```
**Old method:**

step = math.ceil(length / num_slices) = 2
start = i * step, end = (i + 1) * step:
output: [0,1] [2,3] [4,5] ... [98,99] -> 50 slices != num_slices

**New method:**

start = i * length // num_slices, end = (i + 1) * length // num_slices:
output: [0] [1] [2] ... [98,99] -> 99 slices

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Use existing tests.

### CODE REVIEW

**Comment by WeichenXu123** [comment]
ok to test

**Comment by WeichenXu123** [comment]
Jenkins, add to whitelist

File: python/pyspark/sql/pandas/conversion.py [hunk 1]
```diff
@@ -404,8 +404,10 @@ def _create_from_pandas_with_arrow(self, pdf, schema, timezone):
                            for t in pdf.dtypes]
 
         # Slice the DataFrame to be batched
-        step = -(-len(pdf) // self.sparkContext.defaultParallelism)  # round int up
-        pdf_slices = (pdf.iloc[start:start + step] for start in range(0, len(pdf), step))
+        length = len(pdf)
+        num_slices = self.sparkContext.defaultParallelism
+        pdf_slices = (pdf.iloc[i * length // num_slices: (i + 1) * length // num_slices]
+                      for i in range(0, num_slices))
```
Comment (WeichenXu123) [inline python/pyspark/sql/pandas/conversion.py:411]: let's use
```
start_end_pairs = ((i * length // num_slices, (i + 1) * length // num_slices) for i in range(0, num_slices))
pdf_slices = (pdf.iloc[start, end] for start, end in start_end_pairs if start < end)
```
to filter out empty slices (which my leads to error).

**Comment by liangz1** [comment]
This is not a bug. Spark will always create `defaultParallelism` partitions; there could be empty partitions. Closing this PR.

**Pull request closed without merging**